### ISSUE TITLE
[persistence] timed-out blocking read_batch size maximisation

### ISSUE THREAD

**Issue by colinsurprenant** [elastic/logstash#5652]
Following up on a discussion with @guyboertje about the timed-out blocking *read_batch* behaviour. 

When doing a read_batch given a timeout and batch size limit, there are two possibilities:

1. Minimize wait time and return as soon as possible with *some* data, potentially issuing a *small* batch in the case where outputs are faster then inputs. Having faster inputs will result in some accumulated backlog in the queue thus a read_batch would return immediately with a full batch. 

2. Maximize the batch size within the given timeout period. The timeout period would be used to its limit to maximize the batch size. 

The second scenario of maximizing the batch size is slightly more complex to implement but would avoid the possible scenario of producing single event batches when outputs are faster than inputs, without any backlog.

Thoughts?

**Comment by andrewvc** [comment]
My vote is that we go for 1. for the simple reason that it isn't tunable and should work well. Making things that currently need tuning not need tuning is always a good thing IMHO.

I don't have any issue with single event batches. I do think we should prepare ourselves for users who have contrary opinions however.

I think its possible that there are output services that would consider Logstash 'unfriendly' for not maximizing batches, but I can't think of any where that would be a practical problem so I'm not going to object on those grounds. If anyone else can, however, we should carefully consider that.

**Comment by colinsurprenant** [comment]
@andrewvc following that logic, do we need timed-out blocking read_batch at all? a simple blocking read_batch does that already: it returns as soon as any data is available, I don't think we need to exit the read_batch at all if there is no data to read?

**Comment by andrewvc** [comment]
@colinsurprenant no, we don't! Another reason to do this :)

**Comment by ph** [comment]
I would also go for option 1, I think its a simple an elegant solution and its also the same/similar strategy that we have today with the `take_batch`.

I am inclined to believe that in a real world usage the batch should tend to be almost always complete. (Am I optimistic?)

Since its an internal details nothing prevent us in a future to support both strategies and expose that as an option.


> @andrewvc following that logic, do we need timed-out blocking read_batch at all? a simple blocking read_batch does that already: it returns as soon as any data is available, I don't think we need to exit the read_batch at all if there is no data to read?

So does that mean a thread could be blocked forever on `read_batch`, what would be the mechanism to unblock it?

**Comment by andrewvc** [comment]
@ph currently threads block forever on the first queue read. We unblock them by pushing a special 'flush' or 'shutdown' event. I assume the new queue will support the same?

**Comment by ph** [comment]
@andrewvc I would assume the same, but was just curious if we changed that semantic to have a back communication channel.

**Comment by ph** [comment]
@andrewvc if we keep the same concept of sending special event down the queue, theses events should never make it to the persistence layer.

**Comment by colinsurprenant** [comment]
This is a very good point about using events for signalling - obviously, as-is, this will not work since these "signals" would be consumed after any previous backlog which we do not want. I will create a specific issue to tackle this problem. The way  I see it, we should probably introduce a new OOB signalling for shutdown and flush.

**Comment by colinsurprenant** [comment]
@ph 
>I am inclined to believe that in a real world usage the batch should tend to be almost always complete. (Am I optimistic?)

Not really, there are basically 2 behaviours for queues:
1- producer (inputs) is faster than consumer: in this case events are accumulated in the queue and batches should always be full
2- consumer (outputs) is faster than producer: in this case the queue will be mostly always empty and thus batches will be small, possibly single item if we don't try to maximize batch size within the timeout period.

So in other words, a "healthy" logstash which doesn't have a backlog and have outputs faster than inputs, we risk having mostly small/single-item batches.

**Comment by colinsurprenant** [comment]
Here's how I see it: given a non-empty queue, processing this same queue using 2 scenarios: a) by setting batch size to 1 to mimic the small batches situation and b) by setting the batch size to a "normal" size, say of 100+ items) we would definitely see that there is a cost associated with having small/single-item batches. It is overall more efficient to process bigger batches.

The thing here is that this situation would happen only when consumers are faster than producers and thus this inefficiency wouldn't really matter since if it induced slower processing, the queue would start to fill up a little bit and the next batch would be bigger so I feel that any performance penalty induced by processing small batches would actually fix itself in a way, if this performance penalty was happening at or around this producer/consumer speed equilibrium.

**Comment by ph** [comment]
@colinsurprenant Thanks for that detailed explanation.

**Comment by guyboertje** [comment]
My 2 cents...

Assumption: There will be no need for `ls-shipper -> redis/kafka -> ls-ingester`.

This means that the input section will have some beats inputs and some other IP based inputs. The event inflow will not be a firehose drain of redis/kafka buffered events (btw, we have not thought about how complex the input instances/threading/codecs/millling might become).

People will tune the inputs (i.e. the number of inputs and how many beats are connected) and the workers + batch size.

We will have 1) `inflow = outflow`; 2) `inflow < outflow`; 3) `inflow > outflow`.

In the case of 3, higher inflow - it would just fill up the disk and apply back-pressure and for a long running LS we would just have many inactive Pages - maybe a fixed size is better in this case.

Given that inflow can vary we may oscillate between 1, 2 and 3.
However 1) equal and 2) higher outflow would mean small batches.

**Bear in mind, we currently set up to do a few threads handling 500 events each, not 500 threads handling a few events each.**

I would prefer have ES bulk index a full batch than one event at a time.

A big benefit in writing our own queue is to optimise it for our needs - imagine if the queue had an `available_size` and a `last written timestamp` i.e. the batch_take, when it gets notified, goes "I see there are N events available but its less than my batch size and the last one was written before i got parked/notified - I'm going to take these 10 and go" or "I see there are not enough events but some growth/activity is happening - I'm going to spin for a bit and recheck".
We can't get this with other queue implementations.

Therefore I tentatively vote for number 2: Maximize batch size.

How I wish we had a simulator/test rig that could vary the inflow, batch count + workers and output latency, dynamically because we could then test the fitness of either approach.

**Comment by colinsurprenant** [comment]
@guyboertje the 1) inflow = outflow never really exist other than as a very punctual transition state. The fact that  inflow < outflow or inflow > outflow is a bit irrelevant, from a queue perspective the only thing which matter is if it is empty or not. In logstash runtime, any of the inflow < outflow or inflow > outflow can happen or transition to, regardless if the queue is empty or not. But from a read_batch perspective, the only thing which matters is if the queue is empty or not.

For me a disk full situation is an exceptional & catastrophic condition. I would instead talk about having a max byte-size queue option (and/or % disk)  and when reaching it, the same backpressure mechanism that we have with a small fixed-size queue would happen.

Your point for the output **bulking** (ES or any other output) is a very interesting point worth exploring. Remember that small batches will happen **only** in an empty queue situation which necessarily mean the outputs are faster than the inputs. a **bulking** output relying on the batch (they should all rely on the batch moving forward and not do their own buffering) will in this case output single item or small bulks. This should not be a problem and if it is and has a performance impact, the same reasoning above holds: if outputting small bulks is less efficient, the queue will start to fill up at which point batches will become bigger or full which in turn will trickle down as bigger or full bulk requests.

This "organic" batch sizing will have the benefit of not adding latency in flowing events to the outputs: if there is only one event in the queue at time of batch_read, then it will be processed immediately. 

For this, I believe option 1 is still better.

**Comment by jordansissel** [comment]
I choose option 1.

**Comment by andrewvc** [comment]
I have a very simple alternate approach that combines 1. and 2. Poll the queue, and if there is not a full batch wait say 100ms and get whatever's new up to the batch size.

100ms is a very short amount of time and this amount of time really wouldn't impact latency for most logstash users in any noticeable sort of way (a GC could easily take > 100ms).

The nice thing about making this a 'hard' 100ms sleep is that the queue doesn't need an elaborate mechanism to wake a consumer earlier than 100ms, and we still get an appreciable amount of bulk optimization.

**Comment by ph** [comment]
I believe the `organic` scenario described by @colinsurprenant is what will happen. 

We could test solution 1 and @andrewvc's proposal without much changes.

**Comment by jordansissel** [comment]
The 'wait hopefully for a full batch' is an optimization effort similar to the TCP Nagle algorithm. Where Nagle operates on byte sizes of TCP packets, ours would act on event counts (or byte size?) of things in a batch.

In the case where producers (inputs) are outpacing consumers (filters/outputs), then we will likely never wait because we'll always have full batches to deliver to consumers. Similarly, with Nagle, a full tcp segment will be sent immediately and not wait. In this scenario, we won't need to wait anyway.

In the case where consumers are faster than producers, or we have events trickling in in such a way that we almost never have a full 'batch' (or full segment, in the TCP/Nagle example), so waiting will add latency with the hope that more data can pile up to be processed at the end of the waiting. With Nagle, there is a (somewhat?) fixed cost associated with a single packet, so it makes sense to try and fill packets fully before sending -- nagle is enabled by default in most default TCP stacks in my experience. However, in our case, a 'batch' doesn't have a fixed cost in many cases, and our complexity scales by the number of events (or total bytes of those events) to be processed. Given this, I believe that a Nagle-like "wait at most N realtime for more data" will not benefit us much. Additionally, the added hopeful-wait time will increase latency. My experience is that most latency-sensitive network applications will disable Nagle (setting TCP_NODELAY on the socket) to trade for lower latency.

**Comment by colinsurprenant** [comment]
@andrewvc your proposal is effectively #2. having a hard, once X ms timeout is not very different than having a "maximize batch size within X ms". In the latter case, we just loop using a new read timeout set to the difference of X ms - the cumulative time we waited so far, until we reach a timeout of zero or we filled the batch...

**Comment by andrewvc** [comment]
@colinsurprenant  it's different because it doesn't require repeated polls, it just basically tries one simple operation (queue peek then pop) twice, rather than executing a more complex one repeatedly (poll with timeout). Implementing a poll with timeout correctly is harder than peek then pop with a simple sleep.

That being said, thinking through it I don't see the value add as being that great. Let's just do 1. and see if any problems occur.

**Comment by jsvd** [comment]
I'm all for option 1, for the stated reasons that, if the consumers are faster than producers, then you want to lower latency as much as possible.

That said, how is this implemented? In order to "return as soon as possible with some data", how is the "soon" and "some" calculated? Inputs can know if the pipeline is not congested, so once they know that, their either ship a batch with 1 element or create a batch with "some data" based on some other cut off heuristic, like "if pipeline not congested, fill batches half way".

Not having these other parameters will mean that, in a flowing pipeline, batches will be largely composed of one event.

**Comment by colinsurprenant** [comment]
@jsvd absolutely no heuristic involved, just simple thread monitor between the read and the write side. if the queue is empty when trying to to read, it will `wait()` for the `notify()` signal from the write side if it wrote elements in an empty queue. this is how it is currently implemented with monitors and simple method level synchronization, see `write(), nonBlockReadBatch(), readBatch()` methods in https://github.com/elastic/logstash/blob/feature/java_persistence/logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java - note that this is a first & easy implementation WRT thread safety using method level synchronization - in the future we'll probably want to introduce finer grained synchronization. 

>Not having these other parameters will mean that, in a flowing pipeline, batches will be largely composed of one event.

Exactly and as I already discussed this in my previous comments, my reasoning is that this will not be a problem because **if** it does result in a negative performance impact, the problem will *fix itself* since the queue will start to fill and result in producing larger batches - all of this without having to introduce any "artificial" latency by trying to optimize batch size within a timeout period (and yes @andrewvc regardless of the implementation details your proposal is effectively the same as 2. since you introduce a fixed delay to try and maximize the batch size).

**Comment by colinsurprenant** [comment]
as a side note, a timed-out `readBatch()` will be required for #5672 - to make sure the `readBatch()` breaks out of it blocking state at some given maximum interval so the loop gets a chance to verify for shutdown and flush signals - but this timeout will not change the behaviour returning immediately with any data when it is available.

**Comment by guyboertje** [comment]
Thinking about this overnight. I vote for the mechanism that does the least intervention.
I think that, with enough sensor data available on the Queue, different Queue Client classes can easily implement different strategies.

**Comment by colinsurprenant** [comment]
closing. going with strategy 1.
